rust/pattern/conjunction.rs

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+use crate::pattern::Pattern;
+use std::fmt;
+use std::fmt::Display;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Conjunction {
+    pub patterns: Vec<Pattern>,
+}
+
+impl Conjunction {
+    pub fn new(patterns: &[Pattern]) -> Conjunction {
+        Conjunction { patterns: patterns.to_vec() }
+    }
+}
+
+impl<T: Into<Pattern>> From<T> for Conjunction {
```

Comment:
1) and 2) enable `typeql_match(var("x").isa("T"))` and `typeql_match([var("x").isa("T"), var("y").isa("T")])`. 3) is there for consistency mostly (accepting an array but not a vector is a bit weird). It was used before I figured out the `array.into_iter()` issue, and it's used in the latest commit (`[Result<T, E>; N]` is a lot simpler to turn into a `Result<Vec<T>, E>` into `Vec<T>`).

Reply:
Take a look at lines 68-77 here, specifically L75. 

When calling `typeql_match([var("x").isa("movie"), var("c").isa("character")])`, we invoke `Conjunction::from<[Result<Variable, ErrorMessage>; 2]>(...)`. From there we want to do one of two things:
1) if any of those `Result`s `is_err`, return the `Err`;
2) otherwise, gather them into a collection of `Variable`s and construct the `Conjunction`.

The crux here is that `collect()` can take an iterator over `Result<T, E>` and produce a `Result<Vec<T>, E>`, which is exactly what we want. All we need beyond that is just a `From<Vec<T>>` that will construct our `Conjunction`